### PR TITLE
Allow truncating task executions

### PR DESCRIPTION
**Pull request by neob91-close**
**IMPORTANT**
This is a breaking change warranting a major version bump and a breaking changes section in the changelog!

---

This PR introduces a way to configure the limit of executions stored at `t:task:<uuid>:executions`.
We still need to know the total number of times the task has run (for retries), so we're storing the total number of executions in  `t:task:<uuid>:executions_count`.

Redis RPUSH used to add executions to the list is replaced with a script called `TRUNCATED_RPUSH` which removes the left elements as needed.

### CODE REVIEW

**Review by nsaje — COMMENTED** [review]
LGTM! let's hear what @thomasst thinks

**Comment by neob91-close** [comment]
I just realised that we could keep the execution counter inside the execution payload itself instead of maintaining a second counter key. I'm putting this on hold until I hear some feedback about this particular idea and the PR in general.

File: tasktiger/task.py [hunk 1]
```diff
@@ -286,30 +298,38 @@ def _move(self, from_state=None, to_state=None, when=None, mode=None):
...
-                    check_states = {}
-                else:
-                    check_states = {ERROR}
-                scripts.delete_if_not_in_zsets(
-                    _key("task", self.id, "executions"),
-                    self.id,
-                    [_key(state, queue) for state in check_states],
-                    client=pipeline,
-                )
+                if from_state != ERROR:
```
Comment (neob91-close) [inline tasktiger/task.py:301]: Just realised that we should probably just delete these keys if `from_state == ERROR`.

**Review by thomasst — COMMENTED** [review]
> I just realised that we could keep the execution counter inside the execution payload itself instead of maintaining a second counter key. I'm putting this on hold until I hear some feedback about this particular idea and the PR in general.

Is `executions_truncation_counter` just needed to report the total number of executions? If so, I think it would make more sense to just call it `executions_count`, since it's more useful to know the total number of executions rather than amount of executions that got truncated.

Should the `max_tracked_executions` be a per-task setting or a global setting in the config? I almost feel like this is something that would be specific to a deployment rather than the task itself. Or are we thinking we just want to apply it to specific tasks?

File: tasktiger/redis_scripts.py [hunk 2]
```diff
@@ -473,13 +490,15 @@ def srem_if_not_exists(self, key, member, other_key, client=None):
             keys=[key, other_key], args=[member], client=client
         )
 
-    def delete_if_not_in_zsets(self, key, member, set_list, client=None):
+    def delete_if_not_in_zsets(self, to_delete, value, zsets, client=None):
```
[the remark this answers is not in the document]
Comment (neob91-close) [reply tasktiger/redis_scripts.py:493]: https://github.com/closeio/tasktiger/issues/249
Got it on my todo-list

**Comment by neob91-close** [comment]
> > I just realised that we could keep the execution counter inside the execution payload itself instead of maintaining a second counter key. I'm putting this on hold until I hear some feedback about this particular idea and the PR in general.
> 
> Is `executions_truncation_counter` just needed to report the total number of executions? If so, I think it would make more sense to just call it `executions_count`, since it's more useful to know the total number of executions rather than amount of executions that got truncated.

Yes. Did it this way, because it felt more normalised this way, but storing the complete count is fine with me too.

> Should the `max_tracked_executions` be a per-task setting or a global setting in the config? I almost feel like this is something that would be specific to a deployment rather than the task itself. Or are we thinking we just want to apply it to specific tasks?

For some tasks (webhook delivery) we should expect a significant number of retries. For other tasks we don't really need/want to truncate executions.

File: tasktiger/redis_scripts.py [hunk 3]
```diff
@@ -506,6 +522,22 @@ def get_expired_tasks(self, key_prefix, time, batch_size, client=None):
         # [queue1, task1, queue2, task2] -> [(queue1, task1), (queue2, task2)]
         return list(zip(result[::2], result[1::2]))
 
+    def truncated_rpush(self, key, element, max_length, client=None):
```
Comment (thomasst) [inline tasktiger/redis_scripts.py:525]: Just use a pipeline with `rpush` and `ltrim -max_length -1`.

Comment (neob91-close) [reply tasktiger/redis_scripts.py]: bfd5b7b6da67beec141dd277f539f96f50b2ce09

File: tests/test_base.py [hunk 4]
```diff
@@ -269,6 +269,28 @@ def test_exception_task(self, store_tracebacks):
...
+        assert _get_stored_executions() == 0
+
+        for __ in range(6):
+            Worker(self.tiger).run(once=True)
+            time.sleep(DELAY)
+
+            Worker(self.tiger).run(once=True)
+
+        assert task.n_executions() == 6
+        assert _get_stored_executions() == max_stored_executions or 6
```
Comment (thomasst) [inline tests/test_base.py:292]: This assert is true, regardless of the values of `_get_stored_executions()` and `max_stored_executions`  (and also not what you intended even if it was parenthesized as `(max_stored_executions or 6)`).

I believe you want this:
```
assert _get_stored_executions() == min(max_stored_executions or 6, 6)
```

**Review by thomasst — APPROVED** [review]
Looks all good except my comment.

File: tests/test_base.py (hunk 4 above)
Comment (neob91-close) [reply tests/test_base.py]: I think it originally made sense, but the lack of parentheses prevented me from noticing when it ceased to make sense.

Fixed, thanks.

**Review by thomasst — APPROVED** [review]

**Pull request merged**